### PR TITLE
async-44: Fix Pixel Shift Bug

### PR DESCRIPTION
**Pull request by pwinston**
# Description
* This PR is really just cleanup, but some nice cleanup related to geom/visual.
* The real fix is in the DZI plugin: https://github.com/manzt/napari-dzi-zarr/pull/5

# Before

pixel-shift-before

# After

pixel-shift-fixed

# DZI Examples

On this site: http://boschproject.org/#/book/
Click to open the viewer, then use devtools->network to find the URL such as:
http://boschproject.org/dzi/65_66_67_MCPVIS.dzi
http://boschproject.org/dzi/13_14_15_MCPVIS.dzi

# Pixel Shift

Looking at this data:
`NAPARI_OCTREE=1 napari http://hyper-resolution.org/dzi/Rijksmuseum/SK-C-5/SK-C-5_VIS_20-um_2019-12-21.dzi`

There is sometimes a slight shift of a pixel or two when we transition from a lower resolution to a higher resolution. You can see the man in white seems to shift left a bit when the higher resolution starts to draw:

pixel-shift

After the above fixes, it seems like our vertex and texture coordinates are all "perfect" in that everything seems exactly what I expect. So I decided to write out the tile data we are using as PNG's. Here is the level 9 corner tile:

tile_9_1_1

Here is the level 8 corner tile:

tile_8_3_2

Doubling the size of the level 9 tile and then putting them together in photoshop we see the pixel shift:

pixel-shift-photoshop

Since this shows up in Photoshop it's tempting to say they didn't downsample things as carefully as they could have. They are off by a pixel or so. And so the shift is in the data and we can't fix that.

But maybe there some convention here based on the level sizes that they are not actually meant to cover the same physical area, as we assume there are and we assume in napari. So the levels themselves don't sit right on top of each other? Which on the surface seems crazy.

Or maybe this issue is a commonly made "mistake" and we have to try to account for it?

# Overlap

The DZI format has an overlap param. The DZI file above looks like:

```
<?xml version="1.0" encoding="UTF-8"?>
<Image xmlns="http://schemas.microsoft.com/deepzoom/2008"
  Format="jpeg"
  Overlap="1"
  TileSize="254"
  >
  <Size 
    Height="193750"
    Width="231250"
  />
</Image>
```

This discusses what overlap is: https://github.com/VoidVolker/MagickSlicer/issues/2#issue-97766225

And this explains why they (optionally) have overlap: https://github.com/libvips/libvips/issues/795#issuecomment-344788597

It's because some rendering engines have cracks between the tiles of they draw non-overlapping tiles jammed up edge-to-edge. 

Napari's DZI plugin handles overlap here:
https://github.com/manzt/napari-dzi-zarr/blob/master/napari_dzi_zarr/store.py#L82-L107

# Type of change
- [x] Bug-fix in experimental

### CODE REVIEW

**Comment by sofroniewn** [comment]
So interesting. I guess we never saw this on the images we artificially downsampled, but I do see it on my pathology data, so it might be more common unfortunately. Note quite sure what to make of this yet

change_levels

I wonder if @manzt has any experience here or ideas about what might be going on

**Comment by pwinston** [comment]
@sofroniewn @manzt I'm wondering if the linked napari-dzi-zarr code handles overlaps correctly. The way I read it, it seems to say the only time there is overlap is if the `arr` size is larger than the `tilesize`. I don't think that's true.

Instead, I think the overlap is a function of the tile's position in the grid. I think edge and corner tiles have overlaps on 2 or 3 sides even if their size is smaller than tilesize. Even much smaller. For example, I think a tiny corner tile has overlap on two sides no matter how small it is.

Basically, I think overlap is a per-edge thing, there is overlap if that edge borders another tile. Since the intent is to prevent cracks, anywhere a tile borders another tile, you want to duplicate one or more rows from that other tile. I don't think the size of the tile is a factor.

I think the code implicitly assumed power-of-two images, where the tiles are all the same size. With non-power-of-two images, the edge and corner tiles are often smaller than full size. I think this could explain the pixel shift, where edge and corner tiles sometimes shift. I'm going to see if I can get local version of this plugin setup and modify the code.

**Comment by manzt** [comment]
> @sofroniewn @manzt I'm wondering if the linked napari-dzi-zarr code handles overlaps correctly. The way I read it, it seems to say the only time there is overlap is if the arr size is larger than the tilesize. I don't think that's true.

There is a _very_ good chance I did this incorrectly. There is a constant overlap defined in the DZI metadata, but it is my understanding that based on the tile's position in the grid (as you said), and only inner tiles overlap on all sides. More on the DZI overlap here: https://github.com/openseadragon/openseadragon/wiki/The-DZI-File-Format#overlap

Root of the issue is probably here: https://github.com/manzt/napari-dzi-zarr/blob/43f691dcf3b0b673c464484dfc63002ee372cde4/napari_dzi_zarr/store.py#L69, where I try to trim the overlap off each tile based on it's position on the grid. Zarr chunks must all be the same size.

I created the plugin in an afternoon, and I'm sorry that it caused some confusion here!

@sofroniewn -- not sure what's going on with the pathology data. Is that also using my plugin, or are the data zarr?

**Comment by sofroniewn** [comment]
The pathology data comes from https://camelyon16.grand-challenge.org/Data/ and I converted from tif to zarr myself, reading it with [openslide](https://openslide.org/). I didn't use the dzi plugin. The funny thing with that gif is that the tiles look like they jiggle around a little, so not sure what's going on there and if it is related or not to the issue with the dzi image.

**Comment by pwinston** [comment]
The DZI plugin issue is fixed above I think. If there is other jittering/etc would need a repro case. Is this data available?

We could add a config or toggle that turns off all the fancy multi-level rending and puts it into a "all or nothing" mode. Where the new level is not displayed until fully available. 

Although today that would be hard to implement. We need a way to load tiles into VRAM that are not yet drawn. Get the entire new level in their then switch at once. 

That's seems like throwing away a lot of fanciness. But being able to toggle the two modes could be useful.

Basically if "all or nothing" mode fixed your issue, you'd have that as a work around. It's also just might be aesthetically nicer in some cases. Just skip the whole transition period. 

Maybe for lower latency data that's better. You can wait and have it swap all at once. But with high latency data maybe you'd rather see the data as it could. We could even auto-detect, switch all at once if the level is loading fast enough. 

But none of that is easy short term. 

And if we can make multi-level simply perfect then maybe we are fine.

**Comment by sofroniewn** [comment]
Ok, that's great that you identified the fix on the DZI plugin side, I suspect that there is some similar error in my pathology dataset and the "raw conversion" and that these problems are more common than one might originally think, but everything is working well on our octree side. My dataset is a bit hard to share, as it is local, but i'll see if i can fix/ repro elsewhere. I'm happy to review this and merge now though

**Review by sofroniewn — APPROVED** [review]
